### PR TITLE
Updated UGDG

### PR DESCRIPTION
**Pull request by krusagiz**
- Modified UG
- Updated DG
- Added AI to DG

### CODE REVIEW

File: docs/DeveloperGuide.adoc [hunk 1]
```diff
@@ -387,15 +387,19 @@ since `**Budget**` is not really related to `**Transaction**` in the real world.
...
+
+We allow the user to add and remove custom `**Tag**` objects in `**Transaction**` entries so that they can categorise and partition the entries to their own liking.
+The following sections describe how this feature is implemented and the design considerations that led up to the solution.
+
 ==== Implementation
 
-The **tag**/**untag** command performs modifications on existing `**Transaction**` entries. Due to polymorphism, both **Expense**` and `**Income**` objects extends the abstract class `**Transaction**`, and are treated the same way in the context of the **tag**/**untag** command.
+The **tag**/**untag** command performs modifications on existing `**Transaction**` entries. Due to polymorphism, both `**Expense**` and `**Income**` objects extends the abstract class `**Transaction**`, and are treated the same way in the context of the **tag**/**untag** command.
 
 .A Class Diagram illustrating the implementation of Tags inside a Transaction
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: A class diagram (no caps) ... `**Tags**` inside a `**Transaction**

File: docs/DeveloperGuide.adoc [hunk 2]
```diff
@@ -387,15 +387,19 @@ since `**Budget**` is not really related to `**Transaction**` in the real world.
...
-image::tag-untag/Polymorphism.png[]
+image::tag-untag/TagPolymorphism.png[]
 
-Due to the usage of  polymorphism, many of the driver functions in `THRIFT` uses `Transaction` for both `Expense`
-and `Income` objects. For example, when tagging a new `Expense`/`Income`, the `TagCommandParser` will treat them as the same object:
+Due to polymorphic nature of `**Transaction**`, many of the driver functions in `**THRIFT**` uses the `**Transaction**` class for both `**Expense**`
+and `**Income**` objects. For example, when tagging a new `**Expense**`/`**Income**` object, the `**TagCommandParser**` will treat them as the same object:
 
 .An Sequence Diagram showing how the TagCommandParser is called
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Sentence casing, as well as `**TagCommandParser**`

File: docs/DeveloperGuide.adoc [hunk 3]
```diff
@@ -404,17 +408,17 @@ When a user enters a **tag**/**untag** command, it is parsed by the respective `
 
 . Missing parameters
 . Incorrect syntax (i.e. missing prefix, if it is required)
-. Illegal values in parameters (i.e. special characters and symbols entered for an integer-only field)
+. Illegal values in parameters (i.e. non-alphanumeric values given for `*Tag*` names)
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: `**Tag**`

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -404,17 +408,17 @@ When a user enters a **tag**/**untag** command, it is parsed by the respective `
...
 . Multiple occurrences of parameters which only expects a single entry
 
 [NOTE]
 If the user input is incorrect due to any of the reasons above, the corresponding usage syntax will be shown.
 
 The following is an example on how a `**Tag**` is added and removed from a `**Transaction**`, with details on the processes in the backend.
 
-Step 1. The user launches the application with data from previous sessions. THRIFT contains only one `**Transaction**` containing no tags and the user is going to perform tagging and un-tagging operations on it.
+Step 1. The user launches the application with data from previous sessions. THRIFT contains two `**Transaction**` objects, the user is going to perform tagging and un-tagging operations on one of them.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: objects **and** the

File: docs/DeveloperGuide.adoc [hunk 5]
```diff
@@ -424,46 +428,53 @@ Step 2. The user executes the command **tag i/1 t/Delicious** to tag the `**Tran
...
 
 * `**Index**` is parsed by `ParserUtil#parseIndex(String)`
 * `**Tag**` strings are parsed by `ParserUtil#parseTags(Collection<String>)` which calls `ParserUtil#parseTag(String)` iteratively for every string in the collection
 
 [NOTE]
-Tags that do not exist inside the specified `**Transaction**` will be ignored, and if that results in no tags being updated, an error will be shown to the user.
+Tags that do not exist inside the specified `**Transaction**` will be ignored, and if that results in no tags being deleted, an error will be shown to the user.
 
-Again, since the user input is valid, a `**UntagCommand**` command object is created and executed. As a result, a copy of the `**Transaction**` object without the specified `**Tag**` replaces the original in the `**TransactionList**` at the same position.
+Again, since the user input is valid, a `**UntagCommand**` command object is created and executed. As a result, a copy of the `**Transaction**` object with the specified `**Tag**` deleted replaces the original in the `**TransactionList**` at the same position.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc:445]: ... with the specified deleted `**Tag**` replaces

File: docs/DeveloperGuide.adoc [hunk 6]
```diff
@@ -424,46 +428,53 @@ Step 2. The user executes the command **tag i/1 t/Delicious** to tag the `**Tran
...
 
-image::tag-untag/TagUntagTransactionWoTag.png[]
+image::tag-untag/TagUntag1.png[]
 
 The following activity diagrams summarises what happens when the user executes a command to **tag** or **untag** a `**Transaction**`.
 
 .Activity Diagram of the Tagging process
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Sentence casing

File: docs/DeveloperGuide.adoc [hunk 7]
```diff
@@ -424,46 +428,53 @@ Step 2. The user executes the command **tag i/1 t/Delicious** to tag the `**Tran
...
 image::tag-untag/TagActivity.png[]
 
-.Activity Diagram of the Un-tagging process
-image::tag-untag/UntagActivity.png[]
+The **Untag** Command follows the same flow except for the following differences:
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Either `**UntagCommad**` or **Untag** command (small letter `c`)

File: docs/DeveloperGuide.adoc [hunk 8]
```diff
@@ -718,6 +732,52 @@ The added data security will assure the user that their data is safe and allows
...
+//tag::ai[]
+=== [Proposed] Artificial Intelligence (AI)
+Artificial intelligence (AI) is the simulation of human intelligence processes by machines, especially computer systems. These processes include learning (the acquisition of information and rules for using the information), reasoning (using rules to reach approximate or definite conclusions) and self-correction.
+We want `**THRIFT**` to possess the ability to make calculated decisions and provide the user with appropriate recommendations under the appropriate situations.
+This way, we can enhance the user experience and make our application more effective.
+
+The following sections will describe how we plan to implement this feature and the possible design considerations.
+
+==== Proposed Implementation
+Adhering to Open-Closed principle, we will not modify the existing architecture but extend it.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Adhering to **the**

File: docs/DeveloperGuide.adoc [hunk 9]
```diff
@@ -718,6 +732,52 @@ The added data security will assure the user that their data is safe and allows
 
 // end::dataencryption[]
 
+//tag::ai[]
+=== [Proposed] Artificial Intelligence (AI)
+Artificial intelligence (AI) is the simulation of human intelligence processes by machines, especially computer systems. These processes include learning (the acquisition of information and rules for using the information), reasoning (using rules to reach approximate or definite conclusions) and self-correction.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Seems a bit too much background information, can consider shortening it

File: docs/DeveloperGuide.adoc [hunk 10]
```diff
@@ -718,6 +732,52 @@ The added data security will assure the user that their data is safe and allows
...
+* `**AI**`: Analyses data from the `**Data**` and perform machine learning on it.
+If the need arises, feedback to the `**Data**` class to morph the data inside.
+Modifies the actions taken by the `**LogicManager**` class based on the machine learning model.
+
+* `**Data**`:  Collects data from `**Thrift**` and formats to it to a form understandable by the `**AI**` class.
+It can be modified by the `**AI**` class in the process of machine learning.
+
+==== Design consideration
+
+*Aspect: Portability of the Learning Model*
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Sentence casing, also `===== Aspect`

File: docs/DeveloperGuide.adoc [hunk 11]
```diff
@@ -718,6 +732,52 @@ The added data security will assure the user that their data is safe and allows
...
+
+The portability of the learning model will determine the complexity of the implementation, and the following designs can be considered:
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: implementation **and** the

File: docs/DeveloperGuide.adoc [hunk 12]
```diff
@@ -718,6 +732,52 @@ The added data security will assure the user that their data is safe and allows
...
+
+* *Alternative 1*: Include the machine learning algorithms and models with the application
+** Pro: Allows the application to be portable and self-contained
+** Con: Results in larger sizes of the application
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Results in **a** larger **size** of the application

Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Is it supposed to be **pros**/**cons**?

File: docs/DeveloperGuide.adoc [hunk 13]
```diff
@@ -718,6 +732,52 @@ The added data security will assure the user that their data is safe and allows
...
+
+* *Alternative 2*: Develop the machine learning algorithms and models as a separate application,
+which can be offloaded to a remote location
+** Pro: Allows the main application to be smaller in size
+** Con: The application is no longer portable and self-contained,
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Same as above

File: docs/DeveloperGuide.adoc [hunk 14]
```diff
@@ -718,6 +732,52 @@ The added data security will assure the user that their data is safe and allows
...
+and have to depend on the availability of the remote application
+
+We believe that both alternatives can be considered as choosing either option brings a different benefit to THRIFT.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: `**THRIFT**`

File: docs/UserGuide.adoc [hunk 15]
```diff
@@ -149,6 +149,8 @@ NOTE: Assume 2nd command occurs after 1st
 [[TagTag]]
 === Tagging a transaction: `tag`
 
+The `tag` command allows you to tag a `Transaction` entry.
```
Comment (limerencee) [inline docs/UserGuide.adoc:152]: `**Transaction**`

**Review by limerencee — CHANGES_REQUESTED** [review]
LGTM after minor changes

File: docs/DeveloperGuide.adoc (hunk 1 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Changed to sentence case.

File: docs/DeveloperGuide.adoc (hunk 3 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 4 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed

File: docs/DeveloperGuide.adoc (hunk 2 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 5 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc:445]: I don't see any issue with my current wording.

File: docs/DeveloperGuide.adoc (hunk 6 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Changed to sentence case.

File: docs/DeveloperGuide.adoc (hunk 7 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 8 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 9 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: The content has been reduced slightly.

File: docs/DeveloperGuide.adoc (hunk 10 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 11 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 12 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: This has been rephrased.

Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 13 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/DeveloperGuide.adoc (hunk 14 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Fixed.

File: docs/UserGuide.adoc (hunk 15 above)
Comment (krusagiz) [reply docs/UserGuide.adoc:152]: Not sure if it has been standardised in the document yet. We check again,

**Review by limerencee — APPROVED** [review]
LGTM after syncing the branch.

**Pull request merged**